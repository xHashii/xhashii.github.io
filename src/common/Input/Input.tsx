import React, {
  useState,
  createRef,
  HTMLAttributes,
  ReactElement,
  FC
} from "react";
import classNames from "classnames";
import { Classes } from "jss";

export interface IInputProps extends HTMLAttributes<HTMLInputElement> {
  classes: Classes<"input" | "root" | "placeholder" | "placeholderActive" | "inputActive" | "inputIcon">;
  placeholder?: string;
  className?: string;
  icon?: ReactElement;
  type?: string;
  inputIsFocused?(arg0: boolean): void;
  value?: any;
}

const Input: FC<IInputProps> = ({
  classes = {},
  placeholder,
  className,
  icon,
  inputIsFocused = () => {},
  value = null,
  ...props
}) => {
  const [isFocused, setIsFocused] = useState(false);
  const input = createRef<HTMLInputElement>();

  return (
    <div className={classNames(classes.root, className)}>
      {placeholder && (
        <span
          data-test-id="placeholder"
          className={classNames(classes.placeholder, {
            [classes.placeholderActive as any]: isFocused
          })}
        >
          {placeholder}
        </span>
      )}
      {icon && React.cloneElement(icon, { className: classes.inputIcon })}
      <input
        ref={input}
        onFocus={() => {
          setIsFocused(true);

          return inputIsFocused && inputIsFocused(true);
        }}
        onBlur={() => {
          if (input.current) {
            setIsFocused(input.current.value.length > 0);

            return inputIsFocused && inputIsFocused(false);
          }
        }}
        className={classNames(classes.input, {
          [classes.inputActive as any]: isFocused
        })}
        value={value}
        {...props}
      />
    </div>
  );
};

export default Input;
