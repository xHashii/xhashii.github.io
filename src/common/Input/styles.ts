import { Theme } from "../../utils/theme";

export default (theme: Theme) => ({
  input: {
    padding: "12px 16px",
    paddingRight: 32,
    background: theme.colorDark,
    color: 'white',
    outline: "none",
    border: `1px solid ${theme.colorDark}`,
    borderRadius: 4,
    fontSize: 16,
    width: "calc(100% - 50px)",
    fontFamily: theme.fontFamily
  },
  root: {
    position: "relative" as "relative",
    margin: "8px 0",
    display: "inline-block"
  },
  placeholder: {
    position: "absolute" as "absolute",
    fontFamily: theme.fontFamily,
    top: 14,
    left: 12,
    userSelect: "none" as "none",
    pointerEvents: "none" as "none",
    color: "#ccc",
    padding: "0 8px",
    fontSize: 16,
    transition: "all .2s ease-in-out",
  },
  placeholderActive: {
    top: -6,
    left: 8,
    fontWeight: "bold" as "bold",
    background: `linear-gradient(#111111 49%, ${theme.colorDark} 50%)`,
    fontSize: 12,
    color: theme.colorActive
  },
  inputActive: {
    border: `1px solid ${theme.colorActive}`
  },
  inputIcon: {
    position: "absolute" as "absolute",
    color: "#ccc",
    top: 12,
    right: 12,
    fontSize: 20
  }
});
