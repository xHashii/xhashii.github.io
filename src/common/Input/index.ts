import withStyles from "react-jss";

import Input from "./Input";
import styles from "./styles";

export default withStyles(styles)(Input);
