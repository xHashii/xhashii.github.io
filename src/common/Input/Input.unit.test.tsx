import React from "react";
import { shallow } from "enzyme";

import Input from "./Input";

const defaultProps = {
  classes: {}
};

describe("The Input instance", () => {
  describe("When called with default props", () => {
    it("should render without throwing an error", () => {
      const wrapper = shallow(<Input {...defaultProps} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when a placeholder is passed", () => {
    it("should render a placeholder", () => {
      const newProps = {
        ...defaultProps,
        placeholder: "test"
      };
      const wrapper = shallow(<Input {...newProps} />);
      const placeholder = wrapper.find({ "data-test-id": "placeholder" });

      expect(placeholder.exists()).toBe(true);
    });
  });
});
