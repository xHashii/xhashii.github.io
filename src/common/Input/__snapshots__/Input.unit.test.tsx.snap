// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Input instance When called with default props should render without throwing an error 1`] = `ShallowWrapper {}`;
